Sources/TuistCore/Graph/GraphTraverser.swift

```diff
@@ -593,6 +612,61 @@ public class GraphTraverser: GraphTraversing {
         return references
     }
 
+    /// Recursively find platform filters within transitive dependencies
+    /// - Parameters:
+    ///   - rootDependency: dependency whose platform filters we need when depending on `transitiveDependency`
+    ///   - transitiveDependency: target dependency
+    /// - Returns: PlatformFilters to apply to transitive dependency or `nil` if the path to a dependency results in a disjoint
+    /// set of platform filters.
+    func platformFilters(from rootDependency: GraphDependency, to transitiveDependency: GraphDependency) -> PlatformFilters? {
```

Comment:
@waltflanagan perhaps the disjoint logic should be handled elsewhere or differently?

I'm still trying to understand the use case for it, mind sharing an example of where this can occur, is it perhaps a user error that will be caught at lint stage?

Some ideas if we do need to handle this:

- At the point we ask for linkable dependencies / embeddable dependencies during that traversal we could accumulate filters and omit dependencies that have disjoint filters?
- Can we make the disjoint case clearer ?`nil` is a bit ambiguous especially within the traversal where it could mean disjoint or possibly none found?

Reply:
the `nil` case represents both that there is a branch of the tree without the target dependency in it or that there was one, but platform filters as such that it would never be used.  e.g. linking against `FirebaseAnalytics` adds a dependency on `FirebaseCore` which conditionally links against `AppKit` when targetting macOS.  If we are only linking against `FirebaseAnalytics` on iOS then the `AppKit` dependency can just be omitted.

These filter rules need to be applied across all dependencies (this is more evident in the next PR where we have SPM creating multiplatform targets and dependencies) and so everything creating a `GraphDependencyReference` needs to check for filter information.  

I'll explore if an `.all` case for filters would make things more clear.  there's potentially we convert `all` to `[]` when we apply to clean things up some